### DISCUSSION TITLE
Working with graphql

### DISCUSSION THREAD

**Discussion by unolee** [adonisjs/core#1266] [Ideas]
Dear,

I hope it provides a guide on how to integrate with graphql.

I would like to extend current adonis services over to graphql also. 
Or, I would like to build api and graphql at the same time.

I found some old articles;
https://scotch.io/tutorials/build-a-graphql-server-with-apollo-server-and-adonisjs
https://www.apollographql.com/docs/apollo-server/v1/servers/adonis/

However, I hope to get an update with adonisjs v5 and apollo v2.

Do you have any plan?

Best regards,
Uno

**Comment by thetutlage** [comment]
Hey @unolee

 
I see that folks from Apollo have opted the middleware approach for supporting framework https://www.apollographql.com/docs/apollo-server/integrations/middleware/.

I personally do not use GraphQL and hence not a great candidate to write a middleware for AdonisJS. Maybe someone from the community can look into creating one and I will be happy to assist.

As far as I know @ammezie did some work on this for v4.

**Reply by ammezie** [reply]
@thetutlage will check it out and find time to work on for v5.

**Reply by unolee** [reply]
@ammezie @thetutlage Thank you so much for your quick response.

**Reply by targos** [reply]
I started working on an Apollo GraphQL provider for Adonis 5 a few months ago: https://github.com/zakodium/adonis-apollo
Will probably come back to it in the next weeks but it's already working in my application :)

**Reply by sebastiandittrich** [reply]
I have built an Apollo implementation lately for one of my adonis applications. It still needs some work for better configuration and stuff like this, but it is working with default values. Even subscriptions are working, but only when not using multiple node processes. 

If someone is interested and/or wants to contribute, PR's are very welcome since I don't know if I have time to implement all those things!
Here is the Github repo: https://github.com/sebastiandittrich/apollo-adonis

**Comment by bingalls** [comment]
@unolee @thetutlage  I created a working graphql sample, using Adonisjs v5 at https://github.com/bingalls/adonisql
I tried the graphql libs by @ammezie & @sebastiandittrich , and left comments on where I got stuck using them. Feel free to convert my repo into an example of how to use your libs.
Based on my detailed README, a few people have been able to reproduce working graphql servers. The minimal example includes a Query & a Mutation with a database.
I hope to get around to polishing my architecture; in the mean time, feel free to submit a PR.
